src/Cache/RedisCacheTagsChecksum.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\redis\Cache\RedisCacheTagsChecksum.
+ */
+
+namespace Drupal\redis\Cache;
+
+use Drupal\Core\Cache\CacheTagsChecksumInterface;
+use Drupal\Core\Cache\CacheTagsInvalidatorInterface;
+use Drupal\redis\AbstractBackend;
+use Drupal\redis\ClientFactory;
+
+/**
+ * Cache tags invalidations checksum implementation that uses redis.
+ */
+class RedisCacheTagsChecksum extends AbstractBackend implements CacheTagsChecksumInterface, CacheTagsInvalidatorInterface {
+
+  /**
+   * Contains already loaded cache invalidations from the database.
+   *
+   * @var array
+   */
+  protected $tagCache = array();
+
+  /**
+   * A list of tags that have already been invalidated in this request.
+   *
+   * Used to prevent the invalidation of the same cache tag multiple times.
+   *
+   * @var array
+   */
+  protected $invalidatedTags = array();
+
+  /**
+   * @var \Redis
+   */
+  protected $client;
+
+  /**
+   * Creates a PHpRedis cache backend.
+   */
+  function __construct(ClientFactory $factory) {
+    parent::__construct();
+    $this->client = $factory->getClient();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function invalidateTags(array $tags) {
+    $multi =$this->client->multi(\Redis::PIPELINE);
+    $execute = FALSE;
+    foreach ($tags as $tag) {
+      // Only invalidate tags once per request unless they are written again.
+      if (isset($this->invalidatedTags[$tag])) {
+        continue;
+      }
+      $this->invalidatedTags[$tag] = TRUE;
+      unset($this->tagCache[$tag]);
+      $multi->incr($this->getTagKey($tag));
+      $execute = TRUE;
+    }
+    if ($execute) {
+      $multi->exec();
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCurrentChecksum(array $tags) {
+    // Remove tags that were already invalidated during this request from the
+    // static caches so that another invalidation can occur later in the same
+    // request. Without that, written cache items would not be invalidated
+    // correctly.
+    foreach ($tags as $tag) {
+      unset($this->invalidatedTags[$tag]);
+    }
+    return $this->calculateChecksum($tags);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isValid($checksum, array $tags) {
+    return $checksum == $this->calculateChecksum($tags);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function calculateChecksum(array $tags) {
+    $checksum = 0;
+
+    $fetch = array_values(array_diff($tags, array_keys($this->tagCache)));
```

Comment:
@Berdir : So what happens when a concurrent process invalidates a tag and you still have it cached in $this->tagCache?